### PR TITLE
Issue #276: Allow providing Checkstyle version in launch.groovy

### PR DESCRIPTION
**Pull request by Vladlis**
#276

### CODE REVIEW

**Comment by romani** [comment]
I am ok with change.

@rnveach , please review.

**Comment by rnveach** [comment]
@Vladlis CI is also failing for travis.

**Comment by Vladlis** [comment]
Updated diff.groovy as well.

For some reason string interpolation doesn't work. please see failed Travis job: https://travis-ci.org/checkstyle/contribution/jobs/334609495#L817 it searches for `com.puppycrawl.tools:checkstyle:jar:$checkstyleVersion`

Also I tried to make a test commit in https://github.com/checkstyle/checkstyle/pull/5495 and it results in Wercker failure because 7.7-SNAPSHOT is not found, and I have no idea why it tries to use 7.7-SNAPSHOT.

**Comment by romani** [comment]
@Vladlis, please squash commits

**Comment by Vladlis** [comment]
I just thought you could help me with the problems I mentioned, so I made a workaround for one of them as a separate commit to make it clear what am I talking about.
This is a workaround: https://github.com/checkstyle/contribution/pull/277/commits/38fefe9024dd27001bfb8b24130aafbd85218518

File: checkstyle-tester/diff.groovy [hunk 1]
```diff
@@ -180,14 +180,10 @@ def generateCheckstyleReport(cfg) {
...
-    def testerCheckstyleVersion = getCheckstyleVersionFromPomXml('./pom.xml', 'checkstyle.version')
-    def checkstyleVersionInLocalRepo = getCheckstyleVersionFromPomXml("$cfg.localGitRepo/pom.xml", 'version')
-    if (testerCheckstyleVersion != checkstyleVersionInLocalRepo) {
-        throw new GroovyRuntimeException("Error: config version mis-match!\nCheckstyle version in tester's pom.xml is $testerCheckstyleVersion\nCheckstyle version in local repo is $checkstyleVersionInLocalRepo")
-    }
+    def checkstyleVersion = getCheckstyleVersionFromPomXml("$cfg.localGitRepo/pom.xml", 'version')
 
     executeCmd("mvn -Pno-validations clean install", cfg.localGitRepo)
-    executeCmd("groovy launch.groovy --listOfProjects $cfg.listOfProjects --config $cfg.checkstyleCfg --ignoreExceptions --ignoreExcludes")
+    executeCmd("groovy launch.groovy --listOfProjects $cfg.listOfProjects --config $cfg.checkstyleCfg --ignoreExceptions --ignoreExcludes --checkstyleVersion" + checkstyleVersion)
```
Comment (rnveach) [inline checkstyle-tester/diff.groovy]: Aren't you missing a space after `checkstyleVersion`?

File: checkstyle-tester/launch.groovy [hunk 2]
```diff
@@ -22,6 +22,8 @@ def getCliOptions(args) {
         l(longOpt: 'listOfProjects', args: 1, required: true, argName: 'path', 'Path to file which contains projects to test on (required)')
         i(longOpt: 'ignoreExceptions', required: false, 'Whether Maven Checkstyle Plugin should ignore exceptions (optional, default is false)')
         g(longOpt: 'ignoreExcludes', required: false, 'Whether to ignore excludes specified in the list of projects (optional, default is false)')
+        cv(longOpt: 'checkstyleVersion', required: false,
```
Comment (rnveach) [inline checkstyle-tester/launch.groovy]: Isn't this missing `args: 1`?

File: checkstyle-tester/diff.groovy (hunk 1 above)
Comment (Vladlis) [reply checkstyle-tester/diff.groovy]: yes, thanks

File: checkstyle-tester/launch.groovy (hunk 2 above)
Comment (Vladlis) [reply checkstyle-tester/launch.groovy]: thanks, fixed

**Comment by romani** [comment]
Travis is failing, please test Travis's command on your local:
```
...
[INFO] --- maven-site-plugin:3.3:site (default-site) @ sample ---
[INFO] configuring report plugin org.apache.maven.plugins:maven-checkstyle-plugin:2.17
[WARNING] The POM for com.puppycrawl.tools:checkstyle:jar:$checkstyleVersion is missing, no dependency information available
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.788 s
[INFO] Finished at: 2018-01-29T18:07:50Z
[INFO] Final Memory: 20M/619M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project sample: failed to get report for org.apache.maven.plugins:maven-checkstyle-plugin: Plugin org.apache.maven.plugins:maven-checkstyle-plugin:2.17 or one of its dependencies could not be resolved: Failure to find com.puppycrawl.tools:checkstyle:jar:$checkstyleVersion in http://sevntu-checkstyle.github.com/sevntu.checkstyle/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of sevntu-maven has elapsed or updates are forced -> [Help 1]

```

**Comment by Vladlis** [comment]
Fixed Travis, but Wercker still has the same error as in  https://github.com/checkstyle/contribution/pull/277#issuecomment-361247158

**Comment by romani** [comment]
Let's merge this first

**Comment by rnveach** [comment]
@romani before we merge I would like a test run with using commits from 2 different checkstyle versions to verify there are no problems in difference report. If we break regression report, then we won't be able to verify changes in main repo until it is fixed.

@Vladlis How about we compare master (base branch) tohttps://github.com/checkstyle/checkstyle/commit/aba2d99cdeed555c09c3e76dac16978d156a40ef (patch branch) using the module `JavadocMethod`.
This is the commit right before https://github.com/checkstyle/checkstyle/commit/7bde98c60af85ab4f46fbd426508fc0d0f1da295 where we updated `JavadocMethod`'s messages.

**Comment by romani** [comment]
@Vladlis , please confirm that commands from wercker.yml works no NoErrors items, and we will merge this PR

**Comment by rnveach** [comment]
@Vladlis There are no conflicts, please rebase on latest master.

**Comment by Vladlis** [comment]
@rnveach @romani 
rebased
> How about we compare master (base branch) to checkstyle/checkstyle@aba2d99 (patch branch) using the module JavadocMethod.

please see https://vladlis.github.io/276/

**Pull request merged**